tests/api/utils/test_definition_rendering.py

```diff
@@ -622,11 +676,371 @@ def test_os_env_and_override_envs_in_version_1():
+                "deploy_root": "output/deploy/",
+                "generated_root": "__generated/",
+                "scratch_stage": "app_src.stage_snowflake_cli_scratch",
+                "source_stage": "app_src.stage",
+                "package": {
+                    "name": "test_app_pkg_username",
+                    "distribution": "internal",
+                },
+                "application": {
+                    "name": "test_app_username",
+                },
+            },
+            "env": ProjectEnvironment(default_env={}, override_env={}),
+        },
+    }
+
+
+@mock.patch.dict(os.environ, {"stage": "invalid stage name"}, clear=True)
+def test_field_with_custom_validation_with_templates_and_invalid_value():
+    definition = {
+        "definition_version": "1.1",
+        "native_app": {
+            "name": "test_app",
+            "artifacts": [],
+            "source_stage": "<% ctx.env.stage %>",
+        },
+    }
+
+    with pytest.raises(SchemaValidationError) as err:
+        render_definition_template(definition, {})
+
+    assert "Incorrect value for source_stage value of native_app" in err.value.message
+
+
+@pytest.mark.parametrize(
+    "na_name, expected_app_name, expected_pkg_name",
+    [
+        # valid unquoted ID
+        ("safe_name", "safe_name_username", "safe_name_pkg_username"),
+        # valid quoted ID with unsafe char
+        ('"unsafe.name"', '"unsafe.name_username"', '"unsafe.name_pkg_username"'),
+        # valid quoted ID with safe char
+        ('"safe_name"', '"safe_name_username"', '"safe_name_pkg_username"'),
+        # valid quoted id with double quotes char
+        ('"name_""_"', '"name_""__username"', '"name_""__pkg_username"'),
+        # unquoted ID with unsafe char
+        ("unsafe.name", '"unsafe.name_username"', '"unsafe.name_pkg_username"'),
+    ],
+)
+@mock.patch.dict(os.environ, {"USER": "username"}, clear=True)
+def test_defaults_native_app_pkg_name(
+    na_name, expected_app_name: str, expected_pkg_name: str
+):
+
+    definition = {
+        "definition_version": "1.1",
+        "native_app": {"name": na_name, "artifacts": []},
+        "env": {
+            "app_reference": "<% ctx.native_app.application.name %>",
+            "pkg_reference": "<% ctx.native_app.package.name %>",
+        },
+    }
+    result = render_definition_template(definition, {})
+    project_context = result.project_context
+    project_definition = result.project_definition
+
+    assert project_definition.native_app.application.name == expected_app_name
+    assert project_definition.native_app.package.name == expected_pkg_name
+
+    env = project_context.get("ctx", {}).get("env", {})
+    assert env.get("app_reference") == expected_app_name
+    assert env.get("pkg_reference") == expected_pkg_name
+
+
+@pytest.mark.parametrize(
+    "definition",
+    [
+        {
+            "definition_version": "1.1",
+            "native_app": {
+                "name": "myapp",
```

Comment:
tests from here below (All the _suffixing tests) can be removed - they're failing now right?